diff -Naur Ice-3.5.1-old/cpp/config/Make.rules Ice-3.5.1-new/cpp/config/Make.rules
--- Ice-3.5.1-old/cpp/config/Make.rules	2013-10-04 08:48:15.000000000 -0700
+++ Ice-3.5.1-new/cpp/config/Make.rules	2015-06-23 14:11:18.196317788 -0700
@@ -212,13 +212,13 @@
 ifeq ($(OPENSSL_LIBS),)
     OPENSSL_LIBS            = $(if $(OPENSSL_HOME),-L$(OPENSSL_HOME)/$(libsubdir)) -lssl -lcrypto
 endif
-OPENSSL_RPATH_LINK      = $(if $(OPENSSL_HOME),$(call rpathlink,$(OPENSSL_HOME)/$(libsubdir)))
+OPENSSL_RPATH_LINK      = $(if $(OPENSSL_HOME),$(call rpathlink,$(OPENSSL_HOME)/$(libsubdir))) $(OPENSSL_LIBS)
 
 BZIP2_FLAGS             = $(if $(BZIP2_HOME),-I$(BZIP2_HOME)/include)
 ifeq ($(BZIP2_LIBS),)
     BZIP2_LIBS              = $(if $(BZIP2_HOME),-L$(BZIP2_HOME)/$(libsubdir)) -lbz2
 endif
-BZIP2_RPATH_LINK        = $(if $(BZIP2_HOME),$(call rpathlink,$(BZIP2_HOME)/$(libsubdir)))
+BZIP2_RPATH_LINK        = $(if $(BZIP2_HOME),$(call rpathlink,$(BZIP2_HOME)/$(libsubdir))) $(BZIP2_LIBS)
 
 ifeq ($(ICONV_LIBS),)
     ICONV_LIBS              = $(if $(ICONV_HOME),-L$(ICONV_HOME)/$(libsubdir)) $(ICONV_LIB)
@@ -229,7 +229,7 @@
     ifeq ($(DB_LIBS),)
       DB_LIBS              = -L$(DB_HOME)/$(libsubdir)$(cpp11suffix) -ldb_cxx
     endif
-    DB_RPATH_LINK        = $(call rpathlink,$(DB_HOME)/$(libsubdir)$(cpp11suffix))
+    DB_RPATH_LINK        = $(call rpathlink,$(DB_HOME)/$(libsubdir)$(cpp11suffix)) $(DB_LIBS)
 else
    ifeq ($(shell if [ -d /usr/include/db53 -a -d /usr/$(libsubdir)/db53 ] ; then echo yes; fi), yes)
       DB_FLAGS          = -I/usr/include/db53
@@ -248,7 +248,7 @@
 ifeq ($(EXPAT_LIBS),)
     EXPAT_LIBS              = $(if $(EXPAT_HOME),-L$(EXPAT_HOME)/$(libsubdir)) -lexpat
 endif
-EXPAT_RPATH_LINK        = $(if $(EXPAT_HOME),$(call rpathlink,$(EXPAT_HOME)/$(libsubdir)))
+EXPAT_RPATH_LINK        = $(if $(EXPAT_HOME),$(call rpathlink,$(EXPAT_HOME)/$(libsubdir))) $(EXPAT_LIBS)
 
 ifeq ($(MCPP_LIBS),)
     ifeq ($(MCPP_HOME),)
@@ -262,7 +262,7 @@
     endif
     MCPP_LIBS              = $(if $(MCPP_HOME),-L$(MCPP_HOME)/$(libsubdir)) -lmcpp
 endif
-MCPP_RPATH_LINK        = $(if $(MCPP_HOME),$(call rpathlink,$(MCPP_HOME)/$(libsubdir)))
+MCPP_RPATH_LINK        = $(if $(MCPP_HOME),$(call rpathlink,$(MCPP_HOME)/$(libsubdir))) $(MCPP_LIBS)
 
 
 ifeq ($(PLATFORM_HAS_READLINE),yes)
@@ -312,11 +312,11 @@
 endif
 
 ifdef ice_src_dist
-    SLICEPARSERLIB	= $(libdir)/$(call mklibfilename,Slice,$(VERSION))
+    SLICEPARSERLIB	= $(libdir)/$(call mklibname,Slice)
     SLICE2CPP		= $(bindir)/slice2cpp
     SLICE2FREEZE	= $(bindir)/slice2freeze
 else
-    SLICEPARSERLIB	= $(ice_dir)/$(libsubdir)$(cpp11suffix)/$(call mklibfilename,Slice,$(VERSION))
+    SLICEPARSERLIB	= $(ice_dir)/$(libsubdir)$(cpp11suffix)/$(call mklibname,Slice)
     SLICE2CPP		= $(ice_dir)/$(binsubdir)$(cpp11suffix)/slice2cpp
     SLICE2FREEZE	= $(ice_dir)/$(binsubdir)$(cpp11suffix)/slice2freeze
 endif
diff -Naur Ice-3.5.1-old/cpp/config/Make.rules.Linux Ice-3.5.1-new/cpp/config/Make.rules.Linux
--- Ice-3.5.1-old/cpp/config/Make.rules.Linux	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/config/Make.rules.Linux	2015-06-23 18:07:14.643493586 -0700
@@ -150,7 +150,7 @@
     BASELIBS		:= $(NPTL_LIB) $(BASELIBS)
 endif
 
-LIBS			= $(BZIP2_RPATH_LINK) -lIce $(BASELIBS)
+LIBS			= $(BZIP2_RPATH_LINK) -ldl -lIce $(BASELIBS)
 
 ICEUTIL_OS_LIBS         = -lrt
 ICE_OS_LIBS             = -ldl
diff -Naur Ice-3.5.1-old/cpp/src/Freeze/Makefile Ice-3.5.1-new/cpp/src/Freeze/Makefile
--- Ice-3.5.1-old/cpp/src/Freeze/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/Freeze/Makefile	2015-06-23 15:32:07.756492208 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,Freeze,$(VERSION))
 SONAME          = $(call mksoname,Freeze,$(SOVERSION))  
-LIBNAME		= $(call mklibname,Freeze)
+LIBNAME		= $(call mksoname,Freeze)
 
 TARGETS		= $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/src/FreezeScript/Makefile Ice-3.5.1-new/cpp/src/FreezeScript/Makefile
--- Ice-3.5.1-old/cpp/src/FreezeScript/Makefile	2013-10-04 08:48:15.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/FreezeScript/Makefile	2015-06-23 14:21:23.726339573 -0700
@@ -50,11 +50,11 @@
 
 $(TRANSFORMDB): $(TRANSFORM_OBJS) $(COMMON_OBJS)
 	rm -f $@
-	$(CXX) $(LDFLAGS) -o $@ $(TRANSFORM_OBJS) $(COMMON_OBJS) -lSlice $(EXPAT_RPATH_LINK) -lIceXML $(DB_RPATH_LINK) -lFreeze $(LIBS) $(DB_LIBS)
+	$(CXX) $(LDFLAGS) -o $@ $(TRANSFORM_OBJS) $(COMMON_OBJS) $(MCPP_RPATH_LINK) -lSlice $(EXPAT_RPATH_LINK) -lIceXML $(DB_RPATH_LINK) -lFreeze $(LIBS) $(DB_LIBS)
 
 $(DUMPDB): $(DUMP_OBJS) $(COMMON_OBJS)
 	rm -f $@
-	$(CXX) $(LDFLAGS) -o $@ $(DUMP_OBJS) $(COMMON_OBJS) -lSlice $(EXPAT_RPATH_LINK) -lIceXML $(DB_RPATH_LINK) -lFreeze $(LIBS) $(DB_LIBS)
+	$(CXX) $(LDFLAGS) -o $@ $(DUMP_OBJS) $(COMMON_OBJS) $(MCPP_RPATH_LINK) -lSlice $(EXPAT_RPATH_LINK) -lIceXML $(DB_RPATH_LINK) -lFreeze $(LIBS) $(DB_LIBS)
 
 install:: all
 	$(call installprogram,$(TRANSFORMDB),$(DESTDIR)$(install_bindir))
diff -Naur Ice-3.5.1-old/cpp/src/Glacier2Lib/Makefile Ice-3.5.1-new/cpp/src/Glacier2Lib/Makefile
--- Ice-3.5.1-old/cpp/src/Glacier2Lib/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/Glacier2Lib/Makefile	2015-06-23 15:28:41.449818141 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,Glacier2,$(VERSION))
 SONAME          = $(call mksoname,Glacier2,$(SOVERSION))  
-LIBNAME		= $(call mklibname,Glacier2)
+LIBNAME		= $(call mksoname,Glacier2)
 
 TARGETS         = $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/src/IceBox/Makefile Ice-3.5.1-new/cpp/src/IceBox/Makefile
--- Ice-3.5.1-old/cpp/src/IceBox/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceBox/Makefile	2015-06-23 15:30:22.219821746 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IceBox,$(VERSION))
 SONAME          = $(call mksoname,IceBox,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceBox)
+LIBNAME		= $(call mksoname,IceBox)
 
 SERVER		= $(bindir)/icebox
 ADMIN		= $(bindir)/iceboxadmin
diff -Naur Ice-3.5.1-old/cpp/src/IceDB/Makefile Ice-3.5.1-new/cpp/src/IceDB/Makefile
--- Ice-3.5.1-old/cpp/src/IceDB/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceDB/Makefile	2015-06-23 15:28:58.633152080 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IceDB,$(VERSION))
 SONAME          = $(call mksoname,IceDB,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceDB)
+LIBNAME		= $(call mksoname,IceDB)
 
 TARGETS		= $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/src/IceGrid/FreezeDB/Makefile Ice-3.5.1-new/cpp/src/IceGrid/FreezeDB/Makefile
--- Ice-3.5.1-old/cpp/src/IceGrid/FreezeDB/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceGrid/FreezeDB/Makefile	2015-06-23 15:30:10.356488000 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IceGridFreezeDB,$(VERSION))
 SONAME          = $(call mksoname,IceGridFreezeDB,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceGridFreezeDB)
+LIBNAME		= $(call mksoname,IceGridFreezeDB)
 
 TARGETS		= $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/src/IceGrid/SqlDB/Makefile Ice-3.5.1-new/cpp/src/IceGrid/SqlDB/Makefile
--- Ice-3.5.1-old/cpp/src/IceGrid/SqlDB/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceGrid/SqlDB/Makefile	2015-06-23 15:29:49.676487253 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IceGridSqlDB,$(VERSION))
 SONAME          = $(call mksoname,IceGridSqlDB,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceGridSqlDB)
+LIBNAME		= $(call mksoname,IceGridSqlDB)
 
 TARGETS		= $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/src/IceGridLib/Makefile Ice-3.5.1-new/cpp/src/IceGridLib/Makefile
--- Ice-3.5.1-old/cpp/src/IceGridLib/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceGridLib/Makefile	2015-06-23 15:30:37.239822288 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IceGrid,$(VERSION))
 SONAME          = $(call mksoname,IceGrid,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceGrid)
+LIBNAME		= $(call mksoname,IceGrid)
 
 LIBTARGETS	= $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 TARGETS         = $(LIBTARGETS) $(NODE_SERVER) $(REGISTRY_SERVER) $(ADMIN)
diff -Naur Ice-3.5.1-old/cpp/src/IcePatch2Lib/Makefile Ice-3.5.1-new/cpp/src/IcePatch2Lib/Makefile
--- Ice-3.5.1-old/cpp/src/IcePatch2Lib/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IcePatch2Lib/Makefile	2015-06-23 15:31:53.976491713 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IcePatch2,$(VERSION))
 SONAME          = $(call mksoname,IcePatch2,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IcePatch2)
+LIBNAME		= $(call mksoname,IcePatch2)
 
 TARGETS         = $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/src/IceStorm/FreezeDB/Makefile Ice-3.5.1-new/cpp/src/IceStorm/FreezeDB/Makefile
--- Ice-3.5.1-old/cpp/src/IceStorm/FreezeDB/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceStorm/FreezeDB/Makefile	2015-06-23 15:31:19.663157136 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IceStormFreezeDB,$(VERSION))
 SONAME          = $(call mksoname,IceStormFreezeDB,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceStormFreezeDB)
+LIBNAME		= $(call mksoname,IceStormFreezeDB)
 
 MIGRATE		= $(bindir)/icestormmigrate
 
diff -Naur Ice-3.5.1-old/cpp/src/IceStorm/Makefile Ice-3.5.1-new/cpp/src/IceStorm/Makefile
--- Ice-3.5.1-old/cpp/src/IceStorm/Makefile	2013-10-04 08:48:16.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceStorm/Makefile	2015-06-23 15:31:38.943157838 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME  = $(call mklibfilename,IceStormService,$(VERSION))
 SONAME       = $(call mksoname,IceStormService,$(SOVERSION))  
-LIBNAME	= $(call mklibname,IceStormService)
+LIBNAME	= $(call mksoname,IceStormService)
 
 ADMIN		= $(bindir)/icestormadmin
 
diff -Naur Ice-3.5.1-old/cpp/src/IceStorm/SqlDB/Makefile Ice-3.5.1-new/cpp/src/IceStorm/SqlDB/Makefile
--- Ice-3.5.1-old/cpp/src/IceStorm/SqlDB/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceStorm/SqlDB/Makefile	2015-06-23 15:31:08.113156724 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IceStormSqlDB,$(VERSION))
 SONAME          = $(call mksoname,IceStormSqlDB,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceStormSqlDB)
+LIBNAME		= $(call mksoname,IceStormSqlDB)
 
 TARGETS		= $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/src/IceStormLib/Makefile Ice-3.5.1-new/cpp/src/IceStormLib/Makefile
--- Ice-3.5.1-old/cpp/src/IceStormLib/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceStormLib/Makefile	2015-06-23 15:29:13.789819289 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME     = $(call mklibfilename,IceStorm,$(VERSION))
 SONAME          = $(call mksoname,IceStorm,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceStorm)
+LIBNAME		= $(call mksoname,IceStorm)
 
 TARGETS		= $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/src/IceXML/Makefile Ice-3.5.1-new/cpp/src/IceXML/Makefile
--- Ice-3.5.1-old/cpp/src/IceXML/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/src/IceXML/Makefile	2015-06-23 15:30:52.223156157 -0700
@@ -11,7 +11,7 @@
  
 LIBFILENAME     = $(call mklibfilename,IceXML,$(VERSION))
 SONAME          = $(call mksoname,IceXML,$(SOVERSION))  
-LIBNAME		= $(call mklibname,IceXML)
+LIBNAME		= $(call mksoname,IceXML)
 
 TARGETS		= $(call mklibtargets,$(libdir)/$(LIBFILENAME),$(libdir)/$(SONAME),$(libdir)/$(LIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/test/Ice/background/Makefile Ice-3.5.1-new/cpp/test/Ice/background/Makefile
--- Ice-3.5.1-old/cpp/test/Ice/background/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/test/Ice/background/Makefile	2015-06-23 15:32:25.003159495 -0700
@@ -14,7 +14,7 @@
 
 TSPFILENAME     = $(call mklibfilename,TestTransport,$(VERSION))
 TSPSONAME	= $(call mksoname,TestTransport,$(SOVERSION)) 
-TSPLIBNAME	= $(call mklibname,TestTransport)
+TSPLIBNAME	= $(call mksoname,TestTransport)
 
 TARGETS		= $(CLIENT) $(SERVER) $(call mklibtargets,$(TSPFILENAME),$(TSPSONAME),$(TSPLIBNAME))
 
diff -Naur Ice-3.5.1-old/cpp/test/Ice/plugin/Makefile Ice-3.5.1-new/cpp/test/Ice/plugin/Makefile
--- Ice-3.5.1-old/cpp/test/Ice/plugin/Makefile	2013-10-04 08:48:14.000000000 -0700
+++ Ice-3.5.1-new/cpp/test/Ice/plugin/Makefile	2015-06-23 15:32:43.823160172 -0700
@@ -11,7 +11,7 @@
 
 LIBFILENAME	= $(call mklibfilename,TestPlugin,$(VERSION))
 SONAME		= $(call mksoname,TestPlugin,$(SOVERSION))
-LIBNAME		= $(call mklibname,TestPlugin)
+LIBNAME		= $(call mksoname,TestPlugin)
 CLIENT		= $(call mktestname,client)
 PLUGINDIR	= plugins
 
